File: base/Sources/ImportFont.ts

///if (is_paint || is_sculpt)

class ImportFont {

	static run = (path: string) => {
		for (let f of project_fonts) {
			if (f.file == path) {
				console_info(strings_info0());
				return;
			}
		}
		let font: g2_font_t = data_get_font(path);
		g2_font_init(font); // Make sure font_ is ready
		let count: i32 = krom_g2_font_count(font.font_);
		let font_slots: SlotFontRaw[] = [];
		for (let i: i32 = 0; i < count; ++i) {
			let ar: string[] = path.split(path_sep);
			let name: string = ar[ar.length - 1];
			let f: g2_font_t = g2_font_clone(font);
			g2_font_set_font_index(f, i);
			let font_slot: SlotFontRaw = SlotFont.create(name, f, path);
			font_slots.push(font_slot);
		}

		let _init = () => {
			for (let f of font_slots) {
				context_raw.font = f;
				project_fonts.push(f);
				UtilRender.make_font_preview();
			}
		}
		app_notify_on_init(_init);

		UIBase.hwnds[tab_area_t.STATUS].redraws = 2;
	}
}

///end

File: paint/sources/import_font.c

#include "global.h"

void import_font_run_on_next_frame(slot_font_t_array_t *font_slots) {
	for (i32 i = 0; i < font_slots->length; ++i) {
		slot_font_t *f    = font_slots->buffer[i];
		context_raw->font = f;
		any_array_push(project_fonts, f);
		util_render_make_font_preview();
	}
}

void import_font_run(char *path) {
	for (i32 i = 0; i < project_fonts->length; ++i) {
		slot_font_t *f = project_fonts->buffer[i];
		if (string_equals(f->file, path)) {
			console_info(strings_asset_already_imported());
			return;
		}
	}

	draw_font_t *font = data_get_font(path);
	draw_font_init(font); // Make sure font_ is ready
	i32                  count      = draw_font_count(font);
	slot_font_t_array_t *font_slots = any_array_create_from_raw((void *[]){}, 0);

	for (i32 i = 0; i < count; ++i) {
		string_t_array_t *ar   = string_split(path, PATH_SEP);
		char         *name = ar->buffer[ar->length - 1];
		draw_font_t      *f    = GC_ALLOC_INIT(draw_font_t, {.buf = font->buf, .index = font->index});
		draw_font_init(f);
		if (!draw_set_font(f, util_render_font_preview_size)) {
			console_error(tr("Error: Failed to read font data"));
			continue;
		}

		slot_font_t *font_slot = slot_font_create(name, f, path);
		any_array_push(font_slots, font_slot);
	}

	sys_notify_on_next_frame(&import_font_run_on_next_frame, font_slots);

	ui_base_hwnds->buffer[TAB_AREA_STATUS]->redraws = 2;
}
